lib/pbench/server/auth/__init__.py

```diff
@@ -0,0 +1,600 @@
+              https://www.rfc-editor.org/rfc/rfc7662.html#section-4
+        Args:
+            token: token value to introspect
+            token_info_uri: token introspection uri
+        Returns:
+            Extracted token information
+            {
+                "aud": "client_id",
+                "email_verified": "true",
+                "expires_in": "3572", # seconds
+                "access_type": "offline",
+                "exp": "1660087224",
+                "azp": "client_id",
+                "scope": "openid email profile",
+                "email": "user@example.com",
+                "sub": "106203657987367771589"
+            }
+        """
+        payload = {
+            "client_id": self.client_id,
+            "client_secret": self.client_secret_key,
+            "token": token,
+        }
+        return self._post(token_info_uri, data=payload).json()
+
+    def token_introspect_offline(
+        self,
+        token: str,
+        key: str,
+        audience: str = "account",
+        algorithms: List[str] = ["RS256"],
+        **kwargs,
+    ) -> JSON:
+        """
+        Utility method to decode access/Id tokens using the public key provided
+        by the identity provider.
+        The introspected JWT token contains the claims specified in
+        https://tools.ietf.org/html/rfc7662
+        Args:
+            token: token value to introspect
+            key: client public key
+            audience: jwt token audience/client, who this token was intended for
+            algorithms: Algorithm with which this JWT token was encoded
+        Returns:
+            Extracted token information
+            {
+                "aud": "client_id",
+                "email_verified": "true",
+                "expires_in": "3572", # seconds
+                "access_type": "offline",
+                "exp": "1660087224",
+                "azp": "client_id",
+                "scope": "openid email profile",
+                "email": "user@example.com",
+                "sub": "106203657987367771589"
+            }
+        """
+        return jwt.decode(
+            token, key, algorithms=algorithms, audience=audience, **kwargs
+        )
+
+    def get_userinfo(self, token: str = None) -> JSON:
+        """
+        The userinfo endpoint returns standard claims about the authenticated
+        user, and is protected by a bearer token.
+        http://openid.net/specs/openid-connect-core-1_0.html#UserInfo
+        Args:
+            token: Valid token to extract the userinfo
+        Returns:
+            Userinfo payload
+            {
+                "family_name": <surname>,
+                "sub": "106203657987367771589",
+                "picture": <URL>,
+                "locale": "en",
+                "email_verified": true,
+                "given_name": <given-name>,
+                "email": <email-address>,
+                "hd": "redhat.com",
```

Comment:
Again, a token describing what `"hd"` *means* would be more useful, because I assume that `"redhat.com"` isn't the only domain allowed here.

Reply:
I _have_ been wondering what `"hd"` is....  (I presume the `d` is "domain", but I'm coming up empty on the `h`.  🙂)

Reply:
I just find out what `h` is. it stands for Hosted Domain.